timeline-chart/src/layer/time-graph-chart-selection-range.ts

```diff
@@ -17,8 +16,8 @@ export class TimeGraphChartSelectionRange extends TimeGraphLayer {
 
     protected updateScaleAndPosition() {
         if (this.unitController.selectionRange && this.selectionRange) {
-            const firstCursorPosition = this.getPixel(this.unitController.selectionRange.start - this.unitController.viewRange.start);
-            const width = this.getPixel(this.unitController.selectionRange.end - this.unitController.selectionRange.start)
+            const firstCursorPosition = this.getWorldPixel(this.unitController.selectionRange.start);
```

Comment:
This only works if the parent container is an instance of TimeGraphViewportContainer and not a just TimeGraphContainer. Shouldn't APIs be defined in TimeGraphContainer and here you use these APIs. TimeGraphViewportContainer should just overwrite the relevant methods, but in this class only calls an API doesn't have care where it's implemented (TimeGraphContainer or TimeGraphViewportContainer)

Reply:
I'm not sure I understand this one.

TimeGraphViewportLayer#getWorldPixel is just an abstraction of getPixel.  I noticed for every call of #getPixel we were doing: `TimeWeWantPixelFor - unitController.viewRange.start`.  (This became `TimeWeWantPixelFor - unitController.worldRange.start` for the viewport implementation).  getWorldPixel's only purpose is to make the code cleaner, so it can be removed without consequence. 

Furthermore:

All layers added to a TimeGraphViewportContainer must be a TimeGraphViewportLayer.  Abstract class TimeGraphChartLayer now extends TimeGraphViewportLayer, and all layers that are added to the ChartContainer and MarkersContainer (timegraph-output-component.tsx @ 504-505) are now ViewportLayers with access to the inherited method.

There is a check at TimeGraphViewportContainer#addLayer @ 29.  If a layer is not a ViewportLayer, an error is thrown.

The only non-viewport layers are the top-axis, vertical, and horizontal scrollbars.

If the design is wrong, or this doesn't answer your question, please let me know.

Reply:
This is comment is related to the comment in the related `theia-trace-extension` PR [1]. It's not clear from the API when to use a TimeGraphViewportLayer and when to use TimeGraphChartLayer, as well that they cannot be mixed. 

Maybe all layers should be viewport layer and the TimeGraphChartLayer is just an internal base class. Or is there a different way? WDYT?

Btw, the documentation [2] should also have a description about the TimeGraphViewportLayer and its usage.

[1] https://github.com/eclipse-cdt-cloud/theia-trace-extension/pull/832#discussion_r986086848
[2] https://github.com/eclipse-cdt-cloud/timeline-chart/blob/master/doc/documentation.md

Reply:
The need to use a viewport container has been removed.